### PR TITLE
METRON-1893 Syslog RFC-3164 parser

### PR DESCRIPTION
**Pull request by ottobackwards**
This PR adds a new parser for Syslog RFC-3164

This PR also moves from simple-syslog-5424 to simple-syslog, which has both RFC parsers in it.  So the Syslog5424Parser is updated as well.

The parsers are refactored to a common base since they share interfaces.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [-] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [-] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [-] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [-] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-parsers/src/main/resources/META-INF/NOTICE [hunk 1]
```diff
@@ -37,4 +37,18 @@ Copyright 2006-2011 Google, Inc.
...
 
+ (ASLv2) simple-syslog-5424
+    The following NOTICE information applies:
 
+        simple-syslog-5424
+        https://github.com/palindromicity/simple-syslog-5424
+
+        Copyright 2018 simple-syslog-5424 authors.
+
+  (ASLv2) simple-syslog-5424
```
Comment (JonZeolla) [inline metron-platform/metron-parsers/src/main/resources/META-INF/NOTICE]: `s/5424/3164/`?

**Comment by nickwallen** [comment]
The actual parsing here is done by an external dependency at https://github.com/palindromicity/simple-syslog, right?  And you are the sole author of that library?  

My only concern is that we need to look at all external dependencies with a critical eye to ensure there is a community backing those.  A community that can provide continued support with bug and security patches.  

This is the same approach we have taken when reviewing our web dependencies recently. We've tried to avoid dependencies that don't have that community backing.  Its difficult to make the case to accept simple-syslog as a dependency, when we have turned down other similar single author web dependencies.

And this is a difficult point for me to make being that the sole author of simple-syslog is none other than @ottobackwards who is such an active and important contributor to Metron.

I am open to a healthy discussion on this.  I could certainly be wrong on this, but I just want to make sure that we are applying the same standards across the project and ensure that the project can outlive any individual contributor.

Other than that, I really like what you've done with this parser @ottobackwards .

For reference, I am referring to the discussion that started on #1219 and spilled over to the mailing list.

**Comment by ottobackwards** [comment]
So given that, you should move this discussion off this PR and to list because it involves a revert on master and removal of already accepted functionality

**Comment by mmiklavc** [comment]
Adding my 2 cents here, but also happy to add these thoughts to a discuss thread if others want it. I think it definitely makes sense to qualify and justify the libraries we pull into Metron. To that end, we also do depend on many libraries that don't have a big community backing. In all cases I'm aware of, those libraries aren't the choice between "existing huge community-backed project and @ottobackwards personal OSS project." They're typically the only solution, or one of only a handful of a niche set of solutions for a particular problem. I think the main discussion point on https://github.com/apache/metron/pull/1219 was that there were other available solutions that solved the problem well without requiring the need for a custom fork. AFA the original PR that introduced the dependency, https://github.com/apache/metron/pull/1175, the only suggestion I would make going forward from a community perspective is to include a simple reason why a dependency has been added, ie "adding this bc there isn't a library that accomplishes this, it's used on multiple projects, etc." Basically, what Otto already described.

TL;DR - I think it's reasonable to request minimal justification for new libraries, and it's something to keep in mind in future PRs. Seems to me this dep is justified as I'm unaware of other alternatives that solve this problem. I am also very much in favor of enabling community members to contribute to, and cross pollinate with, other projects as Otto has done here.

**Comment by ottobackwards** [comment]
Now that the parser split pr has landed, I have had to deconflict and refactor this.  I have moved the syslog parsers into the parsing-common module, as they are designed to be the base of other parsers, similar to grok and the others.  They parse the wrapper not the message.

For example, I could ( and may ) re-write the CEF, ASA and other processors on top of the syslog processors.

**Comment by nickwallen** [comment]
On the dependency topic...

I am good with depending on simple-syslog assuming there are no other alternatives.  I just want to be sure we're applying the same standards across the project when we look at dependencies.

I do agree that we should provide some minimal justification for new dependencies as @mmiklavc suggested. Going forward, if I introduce a new dependency, I will probably include something like... "Heads up.  There is a new dependency X. I looked at these 3 alternatives with these pros and cons.  I chose X because...".

 And just to be clear, I am not saying that anything wrong was done on #1175.  This is a good contribution to the project.

**Comment by nickwallen** [comment]
Actually, I just noticed that the timestamp that was appended to the 'good' message in my example above is not epoch millis?  Should that be epoch millis?
```
[Stellar]>>> PARSER_PARSE(p, msg)
[
  {
    "syslog.header.hostName": "10.22.8.216",
    "original_string": "<167>Jan  5 08:52:35 10.22.8.216 %ASA-7-609001: Built local-host inside:10.22.8.205",
    "syslog.header.facility": "20",
    "guid": "f9201e92-4c35-4391-95a4-e1d0cd54cdd7",
    "syslog.header.timestamp": "Jan  5 08:52:35",
    "syslog.message": "%ASA-7-609001: Built local-host inside:10.22.8.205",
    "syslog.header.pri": "167",
    "syslog.header.severity": "7",
    "timestamp": "Jan  5 08:52:35",
    "source.type": "syslog"
  }
]
```

File: metron-platform/metron-parsing/metron-parsers-common/src/main/java/org/apache/metron/parsers/syslog/Syslog3164Parser.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.io.Serializable;
+import java.util.EnumSet;
+import java.util.Map;
+
+
+/**
+ * Parser for RFC 3164 messages.
+ */
+public class Syslog3164Parser extends BaseSyslogParser implements Serializable {
```
[the remark this answers is not in the document]
Comment (ottobackwards) [reply metron-platform/metron-parsing/metron-parsers-common/src/main/java/org/apache/metron/parsers/syslog/Syslog3164Parser.java]: I expect the syslog parsers to be a common base for other processors, which are 'wrapped' in syslog.  like the asa and cef.  As such, the syslog capability is a base parser functionality imo.

**Comment by ottobackwards** [comment]
So with regards to the timestamp.  Simple-Syslog doesn't convert the time stamp, that is left to the caller, and that is something that I didn't handle.

I will update the parsers to work like the ASA parser.
And let you know when it is read.

**Comment by nickwallen** [comment]
+1 Looks great.
